File: 백준/Silver/2178. 미로 탐색/미로 탐색.js
let fs = require("fs");
let _ = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

let [_N, ...arr] = _;

const [N, M] = _N.split(" ");
arr = arr.map((v) => v.split("").map(Number));
let dx = [-1, 0, 1, 0];
let dy = [0, -1, 0, 1];

let queue = [];
let dist = Array.from({ length: N }, () => Array.from({ length: M }).fill(1));
arr[0][0] = 0;
queue.push([0, 0]);

while (queue.length) {
  let [x, y] = queue.shift();
  if (x === N - 1 && y === M - 1) {
    break;
  }
  for (let i = 0; i < 4; i++) {
    let nx = dx[i] + x;
    let ny = dy[i] + y;
    if (nx >= 0 && nx < N && ny >= 0 && ny < M && arr[nx][ny]) {
      arr[nx][ny] = 0;
      dist[nx][ny] = dist[x][y] + 1;
      queue.push([nx, ny]);
    }
  }
}
console.log(dist[N - 1][M - 1]);

File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
import sys
import collections
import math
from collections import deque
import copy
import bisect
import itertools
import heapq

#sys.stdin = open("input.txt", "r")

input = sys.stdin.readline

N,M = map(int,input().split())
board = []
visited=[[0 for _ in range(M)] for _ in range(N)]
dx = [-1,0,1,0]
dy = [0,1,0,-1]
for _ in range(N):
    board.append(list(map(int,input().rstrip())))

queue = deque()
queue.append([0,0,1])
board[0][0] = 0

while queue:
    [x,y,dist] = queue.popleft()
    if x == N-1 and y == M-1:
        print(dist)
        break
    for i in range(4):
        nx = x+dx[i]
        ny = y+dy[i]
        if nx < 0 or nx>=N or ny<0 or ny>=M:
            continue
        if board[nx][ny] == 1:
            board[nx][ny] = 0
            queue.append([nx,ny,dist+1])
